src/Carcassonne/Game/Game.cpp

```diff
@@ -205,28 +185,161 @@ bool Game::is_monastery_completed(int x, int y) noexcept {
 void Game::on_monastery_completed(int x, int y, Player player) {
    m_scores.add_points(player, 9);
    std::erase_if(m_figures,
-                 [&x, &y](const Figure &f) {
-                    return f.x == x + 0.5 && f.y == y + 0.5;
+                 [&x, &y, this](const Figure &f) {
+                    if (f.tile_x == x && f.tile_y == y) {
+                       ++m_figure_count[static_cast<mb::size>(f.player)];
+                       return true;
+                    }
+                    return false;
                  });
 }
 
 const ScoreBoard &Game::scores() const noexcept {
    return m_scores;
 }
 
-bool Game::is_town_field_connected(Edge town, Edge field) const noexcept {
-   return std::find(m_towns.cbegin(), m_towns.cend(), std::make_pair(m_groups.group_of(town), m_groups.group_of(field))) != m_towns.end();
+bool Game::is_town_field_connected(Edge town, Edge field) noexcept {
+   std::transform(m_towns.begin(), m_towns.end(), m_towns.begin(), [this](std::pair<Edge, Edge> &pair) {
+      return std::make_pair(static_cast<Edge>(m_groups.group_of(pair.first)), static_cast<Edge>(m_groups.group_of(pair.second)));
+   });
+   return std::find(m_towns.cbegin(), m_towns.cend(), std::make_pair(static_cast<Edge>(m_groups.group_of(town)), static_cast<Edge>(m_groups.group_of(field)))) != m_towns.end();
 }
 
 void Game::draw_tiles() {
    TileType tt = 0;
-   for(const auto& tile : g_tiles) {
-      for(size_t i = 0; i < tile.amount; i++) {
+   for (const auto &tile : g_tiles) {
+      for (size_t i = 0; i < tile.amount; i++) {
          m_tile_set.push_back(static_cast<TileType>(tt));
       }
       tt++;
    }
    std::shuffle(m_tile_set.begin(), m_tile_set.end(), m_random_generator);
 }
 
+void Game::on_next_move(std::function<void(IGame &, Player)> callback) noexcept {
+   m_next_move_callbacks.push_back(callback);
+}
+
+std::vector<Direction> Game::figure_placements(int x, int y) const noexcept {
+   std::vector<Direction> result;
+   result.reserve(8);
```

Comment:
tutaj chyba chciałeś dać 13? Middle+4*Side+8*Corner

Reply:
W większości przypadków będzie mniej niż 8, klasztor z drogą chyba ma 10 i to jest maks. Dla tego przypadku można poświęcić już tą drugą alokację. Ale można pomyśleć, żeby w ogóle nie korzystać wektora tutaj.